### PR TITLE
full_position: use ring order for partition key comparison

### PR DESCRIPTION
**Pull request by denesb**
full_position::cmp is called from storage_proxy to find the minimum last_position across replicas in the EMPTY_REPLICA_PAGES paging path. It was using partition_key::tri_compare (type-aware) instead of ring
order (token first, then legacy byte comparison).

The bug is benign, the comparator is not actually used currently because a tombstones are included in the digest. If the different replicas have different tombstones, a read repair will be triggered.

Fixes: SCYLLADB-2306

Backport: benign issue, no backport needed

### CODE REVIEW

**Comment by avikivity** [comment]
Where is full_position comparison used?

**Comment by avikivity** [comment]
> Where is full_position comparison used?

That is, can we actually construct a case where wrong data is returned?

**Comment by avikivity** [comment]
It was introduced in 3131cbea629d8ac416db592ccfb7a9f55e90c48a, so I guess any paging? But doesn't paging_state contain the token?

**Comment by avikivity** [comment]
```c++
namespace service {
namespace pager {
class paging_state {
    partition_key get_partition_key();
    std::optional<clustering_key> get_clustering_key();
    uint32_t get_remaining_low_bits();
    query_id get_query_uuid() [[version 2.2]] = query_id::create_null_id();
    std::unordered_map<dht::token_range, std::vector<locator::host_id>> get_last_replicas() [[version 2.2]] = std::unordered_map<dht::token_range, std::vector<locator::host_id>>();
    std::optional<db::read_repair_decision> get_query_read_repair_decision() [[version 2.3]] = std::nullopt;
    uint32_t get_rows_fetched_for_last_partition_low_bits() [[version 3.1]] = 0;
    uint32_t get_remaining_high_bits() [[version 4.3]] = 0;
    uint32_t get_rows_fetched_for_last_partition_high_bits() [[version 4.3]] = 0;
    bound_weight get_clustering_key_weight() [[version 5.1]] = bound_weight::equal;
    partition_region get_partition_region() [[version 5.1]] = partition_region::clustered;
};
}
}
```

**Comment by denesb** [comment]
New in v2: added a unit test which demonstrates the wrong `full_position::cmp()` causing a partition to be missing from the query result. *However*, due to another bug in the code, the wrong order cannot be exploited today. I had to fix this other bug to make the test fail.
The "other" bug is this: the `query::result::builder` included the keys of all partitions in the hash, including those which are completely dead. So the vulnerable code path in storage proxy would never trigger, instead a read repair would be triggered.
I fixed this "bug", but I'm not sure about it: on the one hand, fixing this bug means that only content which is actually transmitted to the coordinator is part of the digest, on the other hand, since purgeable tombstones are not included anyway, the digest differences mean real differences  -- a tombstone is missing on one of the nodes, except if there is a timing edge case.

**Comment by denesb** [comment]
This bug cannot lead to real problems in the field, so I will downgrade the priority to P1.

**Comment by denesb** [comment]
New in v2.1: fix race in test between error injection and background replica write.

**Comment by denesb** [comment]
Don't merge this, I need to take another look at the test and maybe drop the digest change.

**Comment by denesb** [comment]
New in v3: 
* the issue turned out to be benign
* drop test and break out cleanup patches into their own PRs
* drop backport labels

**Comment by denesb** [comment]
@avikivity ping.

File: keys/full_position.hh [hunk 1]
```diff
@@ -33,8 +35,12 @@ struct full_position {
...
 
     static std::strong_ordering cmp(const schema& s, const full_position& a, const full_position& b) {
-        partition_key::tri_compare pk_cmp(s);
-        if (auto res = pk_cmp(a.partition, b.partition); res != 0) {
+        // Use ring order (token first, then legacy raw-byte compare) to match the physical
+        // SSTable ordering. Using the type-aware partition_key::tri_compare here would give
+        // wrong results for key types (e.g. timeuuid) where semantic order differs from the
+        // raw-byte order used on disk, potentially causing paging cursors to be advanced past
+        // partitions that haven't been returned yet.
+        if (auto res = dht::ring_position_tri_compare(s, dht::decorate_key(s, a.partition), dht::decorate_key(s, b.partition)); res != 0) {
```
Comment (avikivity) [inline keys/full_position.hh:43]: This is copying the partition key. We need a decorate_key_as_view when the caller promises not to deallocate the partition key.

I guess here it doesn't matter since it will only be used on digest mismatch, yes?

Comment (denesb) [reply keys/full_position.hh:43]: Actually, this is only used when digests match:
https://github.com/scylladb/scylladb/blob/9b22815c6f6b778d80a798b24d15c2485781f28e/service/storage_proxy.cc#L6003-L6010

It was added for the case when some replica has more tombstones and stops earlier. But this would trigger a digest mismatch, so I think this code is never actually used.
I should just drop all this, the hard part is coming up with a test suite that gives me confidence that I didn't miss some use case.

Comment (avikivity) [reply keys/full_position.hh:43]: Ah, so it's only invoked when paging happens. That moves it off the fast path (well paging is in the data path, but anything that happens per-page is amortized by all the rows in the page).

Comment (denesb) [reply keys/full_position.hh:43]: I see you queued this. I'm looking into removing this altogether. But in the meanwhile this PR unlocks other work to remove `partition_key::tri_compare()`, so no more accidental uses can pop up.

Comment (denesb) [reply keys/full_position.hh:43]: I cannot drop this, after adding an exploratory unit test, I realized this is still needed, but in practice will only compare clustering positions within the same partition. I will send a follow-up to eliminate the copies.

Comment (denesb) [reply keys/full_position.hh:43]: I also found SCYLLADB-3527, a very situational apparent data loss bug.
